staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+// ready is a condition variable that represent readiness and
+// allow callers to wait on its status, returning an error with the reason.
+// It allows to use context cancellation and to be stopped,
+// on both cases it stops waiting raising the corresponding error.
+type ready struct {
+	mu      sync.Mutex
+	readyCh chan struct{}
+	stopCh  chan struct{}
+}
+
+// newReady creates a ready condition variable that is not stopped and not ready
+func newReady() *ready {
+	return &ready{
+		readyCh: make(chan struct{}), // is ready if it does not exist
+		stopCh:  make(chan struct{}), // is stopped if it does not exist
+	}
+}
+
+// wait blocks until:
+// - it is ready and returns nil
+// - it is stopped and returns an error
+// - context is cancelled and returns error
+func (r *ready) wait(ctx context.Context) error {
+	select {
+	// if it is ready but context is cancelled
+	// context cancellation fails fast.
+	case <-ctx.Done():
```

Comment:
testcase `TestCacheWatcherContextIfReady` 
https://github.com/kubernetes/kubernetes/pull/108414/files#diff-aed72203cf711bc5f518b68da5f6131358a0a79139efcbfa675f991951712168R604

without prioritization the test is flaky

Reply:
agree, is not completely race free, is just solving the race in case the signal are set before entering the wait(), but is just making the behavior more predictable, without the prio is totally random https://go.dev/play/p/QX7RljvRwyu

Besides it races, this is just short circuiting, if not, the code later has to honor the context